Additional transport using Amazon SNS and SQS

gertjvr opened the pull request:
I have started working on an additional transport using Amazon SNS (topic/exchange) and SQS (queue), its a combination of ActiveMQ and RabbitMQ transport,

Still early days, but would to get some early feedback.

![image](https://user-images.githubusercontent.com/1691062/43045138-2ce3bce2-8df6-11e8-9d89-75f18ba91b97.png)

## Testing locally (need docker environment)
- `git clone https://github.com/localstack/localstack`
- `docker-compose up -d`

## Testing with AWS
- Change your aws credentials in [here](https://github.com/MassTransit/MassTransit/pull/1166/files#diff-b5c51c603d76895f530f9711ea3dab78R121)
### Minimum IAM permissions needed
 - **SNS** `CreateTopic`, `DeleteTopic`, `Subscribe`, `Publish`
 - **SQS** `CreateQueue`, `DeleteQueue`, `SendMessage`, `ReceiveMessage`, `DeleteMessage`

## Configure_Specs are green 

<img width="709" alt="screen shot 2018-07-26 at 8 00 33 am" src="https://user-images.githubusercontent.com/1691062/43230022-230cdeae-90aa-11e8-80d0-f849f816f28a.png">

## ErrorQueue_Specs needs some work.

gtourkas commented:
The lack of SNS support for FIFO queues is kind of a showstopper.

gertjvr commented:
@gtourkas Depends on your workload, ordered messages aren't general supported in message brokers like rabbitmq or azure service bus.

gtourkas commented:
@gertjvr can't say about azure service bus but rabbitmq guarantees ordering of messages via the same (single) connection->channel->exchange->queue->channel->connection path

but indeed many, if not most, workloads do not require ordering of messages 

i did some research on this in the past and indeed the topology to go for pub/sub is SNS topic -> SQS queue 

for an active AWS account you can register with your own Amazon (root) account and create an IAM user play around

SQS and SNS are practically free (first 1 million SQS msgs are free, SNS pushing to SQS has no cost) for experimentation; you can always set a budget alarm for peace of mind

phatboyg commented:
Are you saying this is ready to merge, I see the WIP tag has been removed.

gertjvr commented:
Think so, like @gtourkas mention won't support sqs fifo or encrypted queues, also haven't used it in anger yet.

waynebrantley commented:
@gertjvr What permissions are required for the IAM user at AWS?
Also, assume you are using long polling for SQS consumer?  Does it reconnect after the timeout, didnt see a test for this.

gertjvr commented:
@waynebrantley I have updated the PR description to include the minimum permissions needed, also yes I have implemented long polling by default, there is a property similar to `PerfetchCount` on other transports called `WaitTimeSeconds` using both will have the effect of how many messages  for or how long to wait.

 [example in the test harness](https://github.com/gertjvr/MassTransit/blob/amazonsqs/src/MassTransit.AmazonSqsTransport/Testing/AmazonSqsTestHarness.cs#L121).

Below example will wait for 10 messages or 60 seconds before the next poll cycle.
```
cfg.ReceiveEndpoint(host, "input-queue", x =>
{
    x.PrefetchCount = 10;
    x.WaitTimeSeconds = 60;
    x.Handler<PingMessage>(async context =>
    {
        await context.Publish(new PongMessage(context.Message.CorrelationId));
    });
});
```

[here](https://github.com/gertjvr/MassTransit/blob/amazonsqs/src/MassTransit.AmazonSqsTransport/Contexts/AmazonSqsModelContext.cs#L119) is the long poll implementation its used inside the receive pipeline _and found out the WaitTimeSeconds and Prefetch can be altered while the bus is running._

gertjvr commented:
@waynebrantley good point. Changed it in the [5df82ea](https://github.com/MassTransit/MassTransit/pull/1166/commits/5df82eaa3a1c05d0efc5238133e42a41f7dcc8f4) commit.

kludgeDan commented:
@gertjvr it's great that you added this. I understand how masstransit works with Rabbit MQ but I wasn't sure what I needed to setup in the AWS console (SNS topics or SQS) to get this started. Could you point me to any literature that could get me started with your SQS/SNS support. Any help would be appreciated and I would be willing to help with documentation to give back.

phatboyg commented:
I'm finally digging into this one, have my AWS setup to test it, and am working through some of the design concerns I have about it. Hope to have it sorted out in the next few days.

gertjvr commented:
@kludgeDan you need to create aws security credentials that have create and read permissions for sqs and sns at least, as sqs transport will create sns topics and create the necessary subscriptions to sqs queues.

AWS Console -> IAM -> Users -> Security Credentials

gertjvr commented:
@phatboyg anything I can help with?

phatboyg commented:
You're welcome to look at the commits so far. I've changed some of the way the topology is configured with SNS (only used for publish), changed to use SQS for send directly to the queue, and changed subscriptions to be Raw so that the message headers are properly transferred to SQS. Still need to add SNS-SNS topic subscriptions to support message interface hierarchy.

waynebrantley commented:
@phatboyg The SNS is  like rabbit exchange and SQS is like a rabbit queue.  I guess the way MassTransit works - it does not really use the exchange parts of rabbit much - and always does a one to one mapping - so maybe sending directly to the queue is ok?

gertjvr commented:
@phatboyg Didn't know about the SNS RawMessage option, how does SNS-SNS topic subscriptions work? Amazon SNS doesn't support SNS subscriptions.

waynebrantley commented:
Seems like without using SNS to SQS - a 'publish' would be impossible (where the message potentially goes to many queues).

pha3l commented:
Hello, can ask about the status of this effort? I see there is a MassTransit.AmazonSQS package available on NuGet.  Is it ready for prime time, or am I better off using a different transport for now, perhaps AmazonMQ via the ActiveMQ transport?

I would be interested in contributing to this if there is a need.

phatboyg commented:
It's been merged for a while, and from what I understand it works well. I've only tested it with limited use cases, but it works for me.

pha3l commented:
@phatboyg Thank you, I will give it a try!

JarrodJ83 commented:
Are there plans to add further documentation/examples for using this transport? I used the unit tests to figure some things out. I currently am able to get it to publish messages vis SNS but my ReceiveEndpoint doesn't ever fire even though messages are making it to the queue I've specified. Any suggestions would be much appreciated. Thanks for all the hard work on this. It seems like it will be perfect for my use case if I can figure out how to get it all setup and working.

phatboyg commented:
I just ran the unit tests for the Amazon SQS transport and published messages are received as expected, same with sent messages.

You can see how the test harness is setup:
https://github.com/MassTransit/MassTransit/blob/develop/src/MassTransit.AmazonSqsTransport/Testing/AmazonSqsTestHarness.cs#L122

There isn't a standalone example of Amazon SQS, but aside from the host configuration, any other sample should be easy to modify.

JarrodJ83 commented:
It seems it was an issue with LocalStack. I switched to a real AWS account it seemed to work fine. Thanks for checking on this. Some additional documentation on how the transport works as far as how it ties to Topics, Subscriptions, and Queues would be great none-the-less. I'm really interested in how to control the names of all of those specifics as our IT group will likely want to include those in their CloudFormation templates rather than having the software manage them. Maybe this type of info is provided by other parts of the documentation in more general terms?

optionis-jx commented:
What format should I expect for the host name? I'm having problems connecting from a local machine to AWS with a host format of "amazonsqs://sqs.{region}.amazonaws.com/"

phatboyg commented:
Pretty sure you can just pass the region name to `.Host`:

https://github.com/MassTransit/MassTransit/blob/develop/src/MassTransit.AmazonSqsTransport.Tests/Configure_Specs.cs#L313

RodicaRadut07 commented:
Hello, I am facing the same issue as @JarrodJ83 mentioned earlier. 
I have setup receive endpoint using a real aws instance which works fine, but when I try to test this locallt using local stack the receive endpoint never fires even thought the data arrives in the correct queue and correct format. 
The test samples for connecting a consumer to localsStack didn't work for me. 
Anyone else came across this ?

rabberbock commented:
@RodicaRadut07 I had trouble setting it up with localstack too, but got it to work with an earlier version of localstack. 0.8.0 running on docker. Maybe you want to give that a try.

Acidfabric commented:
any examples how it should work with aws lambda?

gertjvr commented:
@Acidfabric missing some context, are you asking if you can get localstack working with lambda or sns-sqs transport to work with lambda functions?

Acidfabric commented:
I would like to use massTransit sqs/sns transport in aws lambda functions.

gertjvr commented:
@Acidfabric that isn't currently possible.

Acidfabric commented:
Ok, thanks for information

phatboyg commented:
So, how can we make this work with Lambda? Because MassTransit works with Azure Functions when using either Service Bus or Event Hub, via the ability to deliver a message directly to the message receiver. Could follow a similar approach for Lambda I would think.

Acidfabric commented:
As far I can see after deployment there is no sns/sqs trigger attached to Lambda functions.

gertjvr commented:
@phatboyg  a similar approach could be possible, with sns -> lambda subscriptions.

mbrookson commented:
Hi. Is there a way to completely disable SNS so only SQS is used? Thanks.

phatboyg commented:
You can't disable it, but if you specify `ConfigureConsumeTopology = false` on receive endpoints, and only use `.Send` it shouldn't do anything with SNS unless a fault occurs, where it will try to publish the `Fault<T>`. Though you can reconfigure the error pipe I suppose.
